onnx/ir.h

```diff
@@ -0,0 +1,1048 @@
+  // We can't return a std::vector<Node*>& because there's no
+  // way to soundly cast to std::vector<const Node*> (an insane
+  // implementation of std::vector could make this representationally
+  // different.)
+  ArrayRef<Value*> inputs() {
+    return inputs_;
+  }
+  ArrayRef<const Value*> inputs() const {
+    // Vectors are not convertible in const-ness of elements, but
+    // raw pointers are.
+    return {inputs_.data(), inputs_.size()};
+  }
+  // NB: This returns an ArrayRef; that means that it will
+  // get invalidated if you resize inputs (e.g., using addInput)
+  // We can't return a std::vector<Node*>& because there's no
+  // way to soundly cast to std::vector<const Node*> (an insane
+  // implementation of std::vector could make this representationally
+  // different.)
+  ArrayRef<Value*> outputs() {
+    return outputs_;
+  }
+  ArrayRef<const Value*> outputs() const {
+    // Vectors are not convertible in const-ness of elements, but
+    // raw pointers are.
+    return {outputs_.data(), outputs_.size()};
+  }
+  bool hasUses() const {
+    for(auto o : outputs()) {
+      if(o->uses().size() > 0)
+        return true;
+    }
+    return false;
+  }
+  void replaceAllUsesWith(Node * n) {
+    ONNX_ASSERT(outputs().size() == n->outputs().size());
+    size_t nOutputs = outputs().size();
+    for(size_t i = 0; i < nOutputs; i++) {
+      outputs()[i]->replaceAllUsesWith(n->outputs()[i]);
+    }
+  }
+  // lots of things like chunk have a single input or singel output, so we have a
+  // helper to make accessing it easier
+  Value * input() {
+    ONNX_ASSERT(inputs_.size() == 1);
+    return inputs_.at(0);
+  }
+  Value * output() {
+    ONNX_ASSERT(outputs_.size() == 1);
+    return outputs_.at(0);
+  }
+  const  Value * input() const {
+    ONNX_ASSERT(inputs_.size() == 1);
+    return inputs_.at(0);
+  }
+  // Access a particular input.  This is a checked index.
+  Value * input(size_t i) {
+    return inputs_.at(i);
+  }
+  const Value * input(size_t i) const {
+    return inputs_.at(i);
+  }
+
+  // Graphs
+
+  // Note [Topological invariant]
+  // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // We always maintain an up-to-date topological ordering of all nodes via
+  // the next()/prev() links.  All transformations to graphs must preserve
+  // this topological ordering: for example, it is only valid to 'addInput'
+  // with an input which is topologically before the current node.
+  //
+  // Usually, it is obvious whether or not topological order is maintained;
+  // for example, if you are adding nodes to the end of the topsort, it's
+  // impossible for them to refer to inputs that are not in the topsort.
+  // If it is not obvious, please comment accordingly.
+
+  // Add 'node' as an input to 'this' at the end of existing
+  // arguments.  Returns the added node for ease of chaining.
+  //
+  // Given:   %3 = f(%1, %2)
+  // Execute: %3.addInput(%4)
+  // Result:  %3 = f(%1, %2, %4)
+  Value* addInput(Value * node) {
```

Comment:
node --> node_output?

So topological order maintained by upper layer? there may be multiple ways of top sort, how to handle?

Reply:
Yes, you're required to maintain topsort whenever you do operations, but the Graph always maintains a topological order (which is not uniquely determined). Sometimes it may be profitable to move nodes around to a different valid topological sort.